File: cpp-eicio/src/eicio/reader.cc
#include <fcntl.h>
#include <string.h>
#include <iostream>

#include "event.h"
#include "reader.h"
#include "writer.h"

#include <google/protobuf/io/gzip_stream.h>
#include <google/protobuf/io/zero_copy_stream_impl.h>

using namespace google::protobuf;

using namespace eicio;

Reader::Reader(int fd, bool gzip) {
    inputStream = NULL;
    fileStream = NULL;

    auto fileStream = new io::FileInputStream(fd);
    fileStream->SetCloseOnDelete(true);
    inputStream = fileStream;

    if (gzip) {
        inputStream = new io::GzipInputStream(inputStream);
        this->fileStream = fileStream;
    }
}

Reader::Reader(std::string filename) {
    inputStream = NULL;
    fileStream = NULL;

    int fd = open(filename.c_str(), O_RDONLY);
    if (fd != -1) {
        auto fileStream = new io::FileInputStream(fd);
        fileStream->SetCloseOnDelete(true);
        inputStream = fileStream;

        string gzipSuffix = ".gz";
        int sfxLength = gzipSuffix.length();
        if (filename.length() > sfxLength) {
            if (filename.compare(filename.length() - sfxLength, sfxLength, gzipSuffix) == 0) {
                inputStream = new io::GzipInputStream(inputStream);
                this->fileStream = fileStream;
            }
        }
    }
}

Reader::~Reader() {
    if (inputStream) delete inputStream;
    if (fileStream) delete fileStream;
}

Event *Reader::Get() {  // TODO: figure out error handling for this
    if (!inputStream) return NULL;
    io::CodedInputStream stream(inputStream);

    uint32 n;
    if ((n = syncToMagic(&stream)) < 4) return NULL;

    uint32 headerSize;
    if (!stream.ReadLittleEndian32(&headerSize)) return NULL;
    uint32 payloadSize;
    if (!stream.ReadLittleEndian32(&payloadSize)) return NULL;

    auto headerLimit = stream.PushLimit(headerSize);
    auto header = new model::EventHeader;
    if (!header->MergeFromCodedStream(&stream) || !stream.ConsumedEntireMessage()) {
        delete header;
        return Get();  // Indefinitely attempt to resync to magic numbers
    }
    stream.PopLimit(headerLimit);

    auto event = new Event;
    event->SetHeader(header);
    auto *payload = (unsigned char *)event->SetPayloadSize(payloadSize);
    if (!stream.ReadRaw(payload, payloadSize)) {
        delete event;
        return NULL;
    }

    return event;
}

model::EventHeader *Reader::GetHeader() {  // TODO: figure out error handling for this
    if (!inputStream) return NULL;
    io::CodedInputStream stream(inputStream);

    uint32 n;
    if ((n = syncToMagic(&stream)) < 4) return NULL;

    uint32 headerSize;
    if (!stream.ReadLittleEndian32(&headerSize)) return NULL;
    uint32 payloadSize;
    if (!stream.ReadLittleEndian32(&payloadSize)) return NULL;

    auto headerLimit = stream.PushLimit(headerSize);
    auto header = new model::EventHeader;
    if (!header->MergeFromCodedStream(&stream) || !stream.ConsumedEntireMessage()) {
        delete header;
        return GetHeader();  // Indefinitely attempt to resync to magic numbers
    }
    stream.PopLimit(headerLimit);

    if (!stream.Skip(payloadSize)) {
        delete header;
        return NULL;
    }

    return header;
}

int Reader::Skip(int nEvents) {
    if (!inputStream) return -1;
    io::CodedInputStream stream(inputStream);

    int nSkipped = 0;
    for (int i = 0; i < nEvents; i++) {
        uint32 n;
        if ((n = syncToMagic(&stream)) < 4) return -1;

        uint32 headerSize;
        if (!stream.ReadLittleEndian32(&headerSize)) return -1;
        uint32 payloadSize;
        if (!stream.ReadLittleEndian32(&payloadSize)) return -1;

        if (!stream.Skip(headerSize + payloadSize)) return -1;

        nSkipped++;
    }

    return nSkipped;
}

uint32 Reader::syncToMagic(io::CodedInputStream *stream) {
    unsigned char num;
    uint32 nRead = 0;

    while (stream->ReadRaw(&num, 1)) {
        nRead++;

        if (num == magicBytes[0]) {
            bool goodSeq = true;

            for (int i = 1; i < 4; i++) {
                if (!stream->ReadRaw(&num, 1)) break;
                nRead++;

                if (num != magicBytes[i]) {
                    goodSeq = false;
                    break;
                }
            }
            if (goodSeq) break;
        }
    }
    return nRead;
}

File: java-eicio/src/main/java/eicio/Reader.java
package eicio;

import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.io.IOException;
import java.lang.Iterable;
import java.util.Iterator;
import java.util.zip.GZIPInputStream;

import com.google.protobuf.InvalidProtocolBufferException;

public class Reader implements Iterable<Event>, Iterator<Event>
{
	InputStream stream = null;
	InputStream subStream = null;

	private Event queuedEvent = null;

	public Reader(String filename)
			throws IOException {
		stream = new FileInputStream(filename);

		if (filename.endsWith(".gz")) {
			subStream = stream;
			this.stream = new GZIPInputStream(subStream);
		}
	}

	public Reader(InputStream stream, boolean compress)
			throws IOException {
		this.stream = stream;

		if (compress) {
			subStream = stream;
			this.stream = new GZIPInputStream(subStream);
		}
	}

	public Reader(InputStream stream)
			throws IOException {
		this(stream, false);
	}

	public void close()
			throws IOException {
		stream.close();

		if (subStream != null) {
			subStream.close();
		}
	}

	public Event get()
			throws IOException,
			InvalidProtocolBufferException {
		if (queuedEvent != null) {
			Event returnEvent = queuedEvent;
			queuedEvent = null;
			return returnEvent;
		}

		int n = this.syncToMagic();
		if (n < 4) return null;

		long headerSize = getUnsignedInt();
		long payloadSize = getUnsignedInt();

		byte[] headerBuf = new byte[(int)headerSize];
		int nRead = getBytes(headerBuf);
		if (nRead != headerSize) return null;
		Model.EventHeader header = Model.EventHeader.parseFrom(headerBuf);

		byte[] payload = new byte[(int)payloadSize];
		nRead = getBytes(payload);
		if (nRead != payloadSize) return null;
		
		Event event = new Event(header, payload);

		return event;
	}

	public Model.EventHeader getHeader()
			throws IOException,
			InvalidProtocolBufferException {
		if (queuedEvent != null) {
			Event returnEvent = queuedEvent;
			queuedEvent = null;
			return returnEvent.header;
		}

		int n = this.syncToMagic();
		if (n < 4) return null;

		long headerSize = getUnsignedInt();
		long payloadSize = getUnsignedInt();

		byte[] headerBuf = new byte[(int)headerSize];
		int nRead = getBytes(headerBuf);
		if (nRead != headerSize) return null;
		Model.EventHeader header = Model.EventHeader.parseFrom(headerBuf);

		skipBytes(payloadSize);

		return header;
	}

	public int skip(int nEvents)
			throws IOException {
		int nSkipped = 0;
		for (int i = 0; i < nEvents; i++) {
			if (queuedEvent != null) {
				queuedEvent = null;
				nSkipped++;
				continue;
			}

			int n = this.syncToMagic();
			if (n < 4) return nSkipped;

			long headerSize = getUnsignedInt();
			long payloadSize = getUnsignedInt();

			skipBytes(headerSize + payloadSize);
			nSkipped++;
		}

		return nSkipped;
	}

	public Iterator<Event> iterator() {
		return this;
	}

	public boolean hasNext() {
		try {
			if ((queuedEvent = get()) != null)
				return true;
		} catch (Throwable e) {
			e.printStackTrace();
		}
		return false;
	}

	public Event next() {
		try {
			return get();
		} catch (Throwable e) {
			e.printStackTrace();
		}
		return null;
	}

	private int getBytes(byte[] buf)
			throws IOException {
		int nRead = 0;
		while (true) {
			int nBytes = stream.read(buf, nRead, buf.length - nRead);
			if (nBytes < 0) return nRead;
			nRead += nBytes;
			if (nRead == buf.length) return nRead;
		}
	}

	private long skipBytes(long n)
			throws IOException {
		long nSkipped = 0;
		while (true) {
			long nBytes = stream.skip(n - nSkipped);
			if (nBytes < 0L) return nSkipped;
			nSkipped += nBytes;
			if (nSkipped == n) return nSkipped;
		}
	}

	private int syncToMagic()
			throws IOException {
		int nRead = 0;
		while (true) {
			int thisInt;
			if ((thisInt = stream.read()) == -1) return -1;
			byte thisByte = (byte)thisInt;
			nRead++;

			if (thisByte == Writer.magicBytes[0]) {
				boolean goodSeq = true;

				for (int i = 1; i < 4; i++) {
					if ((thisInt = stream.read()) == -1) return -1;
					thisByte = (byte)thisInt;
					nRead++;

					if (thisByte != Writer.magicBytes[i]) {
						goodSeq = false;
						break;
					}
				}

				if (goodSeq) break;
			}
		}

		return nRead;
	}

	private long getUnsignedInt()
			throws IOException {
		final int nBytes = 4;
		int[] theseBytes = new int[nBytes];
		for (int i = 0; i < nBytes; i++) {
			theseBytes[i] = stream.read();
			if (theseBytes[i] == -1) return -1L;
		}

		long num = 0;
		for (int i = 0; i < nBytes; i++)
			num |= (theseBytes[i] & 0xFFL) << (i * 8);
		return num;
	}
}
